os/win/dnd.cpp

```diff
@@ -0,0 +1,335 @@
+    op |= os::DropOperation::Move;
+
+  if (pdwEffect & DROPEFFECT_LINK)
+    op |= os::DropOperation::Link;
+
+  return static_cast<os::DropOperation>(op);
+}
+
+
+gfx::Point drag_position(HWND hwnd, POINTL& pt)
+{
+  ScreenToClient(hwnd, (LPPOINT) &pt);
+  return gfx::Point(pt.x, pt.y);
+}
+
+namespace os {
+
+// HGLOBAL Locking/Unlocking wrapper
+template <typename T>
+class GLock {
+public:
+  GLock() = delete;
+  GLock(const GLock&) = delete;
+  GLock(HGLOBAL hglobal) : m_hmem(hglobal) {
+    m_data = static_cast<T>(GlobalLock(m_hmem));
+  }
+
+  ~GLock() {
+    GlobalUnlock(m_hmem);
+  }
+
+  operator HGLOBAL() {
+    return m_hmem;
+  }
+
+  operator T () {
+    return m_data;
+  }
+
+  SIZE_T size() {
+    return GlobalSize(m_hmem);
+  }
+
+private:
+  HGLOBAL m_hmem;
+  T m_data;
+};
+
+base::paths DragDataProviderWin::getPaths()
+{
+  base::paths files;
+  FORMATETC fmt;
+  fmt.cfFormat = CF_HDROP;
+  fmt.ptd = nullptr;
+  fmt.dwAspect = DVASPECT_CONTENT;
+  fmt.lindex = -1;
+  fmt.tymed = TYMED::TYMED_HGLOBAL;
+  STGMEDIUM medium;
+  if (m_data->GetData(&fmt, &medium) == S_OK) {
+    {
+      GLock<HDROP> hdrop(medium.hGlobal);
+      if (static_cast<HDROP>(hdrop)) {
+        int count = DragQueryFile(hdrop, 0xFFFFFFFF, nullptr, 0);
+        for (int index = 0; index < count; ++index) {
+          int length = DragQueryFile(hdrop, index, nullptr, 0);
+          if (length > 0) {
+            std::vector<TCHAR> str(length + 1);
+            DragQueryFile(hdrop, index, &str[0], str.size());
+            files.push_back(base::to_utf8(&str[0]));
+          }
+        }
+      }
+    }
+    ReleaseStgMedium(&medium);
+  }
+  return files;
+}
+
+SurfaceRef DragDataProviderWin::getImage()
+{
+  SurfaceRef surface = nullptr;
+  clip::image_spec spec;
+  clip::image img;
+  SurfaceFormatData sfd;
+
+  STGMEDIUM medium;
+  medium.hGlobal = nullptr;
+  FORMATETC fmt;
+  fmt.ptd = nullptr;
+  fmt.dwAspect = DVASPECT_CONTENT;
+  fmt.lindex = -1;
+  fmt.tymed = TYMED_HGLOBAL;
+
+  UINT png_format = RegisterClipboardFormatA("PNG");
+  if (png_format) {
+    fmt.cfFormat = png_format;
+    if (m_data->GetData(&fmt, &medium) == S_OK) {
+      GLock<uint8_t*> png_handle(medium.hGlobal);
+      if (clip::win::read_png(png_handle, png_handle.size(), &img, nullptr)) {
+        spec = img.spec();
+        goto makeSurface;
+      }
+    }
+  }
+
+  fmt.cfFormat = CF_DIBV5;
+  if (m_data->GetData(&fmt, &medium) == S_OK) {
+    GLock<BITMAPV5HEADER*> b5(medium.hGlobal);
+    clip::win::BitmapInfo bi(b5);
+    if (bi.to_image(img)) {
+      spec = img.spec();
+      goto makeSurface;
+    }
+  }
+
+  fmt.cfFormat = CF_DIB;
+  if (m_data->GetData(&fmt, &medium) == S_OK) {
+    GLock<BITMAPINFO*> hbi(medium.hGlobal);
+    clip::win::BitmapInfo bi(hbi);
+    if (bi.to_image(img)) {
+      spec = img.spec();
+      goto makeSurface;
+    }
+  }
+
+  // No suitable image format found, release medium and return.
+  goto releaseMedium;
```

Comment:
This indicates that probably we need a wrapper (like `GLock`) to automatically call `ReleaseStgMedium` in its dtor and avoid two goto labels in the same function.

Probably we can reorganize the logic of this function, e.g. creating other private `DragDataProviderWin` function.